assembly/src/lib.rs

```diff
@@ -43,6 +45,63 @@ const MODULE_PATH_DELIM: &str = "::";
 type ProcMap = BTreeMap<String, Procedure>;
 type ModuleMap = BTreeMap<String, ProcMap>;
 
+// MODULE PROVIDER
+// ================================================================================================
+
+/// The module provider is now a simplified version of a module cache. It is expected to evolve to
+/// a general solution for the module lookup
+///
+/// TODO compute a procedure index deterministically from a module path & procedure label. The
+/// initial expected layout for the index is `[u8; 20]`. The module provider should map procedures
+/// from these indexes, and this is expected to result in code simplification + optimization for
+/// the compiler.
+pub trait ModuleProvider {
+    /// Fetch source contents provided a module path
+    fn get_source(&self, path: &str) -> Option<&str>;
```

Comment:
nit: should we rename this to `get_module`? In the future, the signature here will probably change to something like:
```Rust
fn get_module(&self, index: [u8; 20]) -> Option<&ModuleAst>;
```

Reply:
The original idea I had in mind is to keep it like `get_source` so we go for a minimal set of changes by reflecting the current behavior of stdlib.

The next step is to introduce the procedure index that will most likely deprecate this `get_source` as it is potentially a leaked internals of the module provider.

I'm adding a comment there explaining this approach